xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc

```diff
@@ -152,99 +159,413 @@ bool IsAlignedSlice(const HloInstruction* slice) {
+  } else {
+    TF_RETURN_IF_ERROR(whileop->parent()->ReplaceInstructionWithDifferentShape(
+        whileop, new_whileop));
+  }
+
+  // Next, lets handle the condition
+  while_cond->ReplaceParameter(0, HloInstruction::CreateParameter(
+                                      0, new_while_init->shape(), "new_param"));
+
+  // Next, lets handle the body
+  HloInstruction* new_body_param = while_body->ReplaceParameter(
+      0,
+      HloInstruction::CreateParameter(0, new_while_init->shape(), "new_param"));
+
+  // Next, update the value of the param inside while op
+  HloInstruction* gte = while_body->AddInstruction(
+      HloInstruction::CreateGetTupleElement(
+          new_body_param, new_while_init->shape().tuple_shapes_size() - 1),
+      "loop_iteration_count");
+  HloInstruction* c1 = while_body->AddInstruction(
+      HloInstruction::CreateConstant(LiteralUtil::CreateR0(indvar_type, 1)),
+      "one");
+  HloInstruction* add = while_body->AddInstruction(
+      HloInstruction::CreateBinary(gte->shape(), HloOpcode::kAdd, gte, c1),
+      "updated_loop_iteration_count");
+  absl::InlinedVector<HloInstruction*, 2> old_return_tuple_operands =
+      while_body->root_instruction()->operands();
+  std::vector<HloInstruction*> new_return_tuple_operands(
+      old_return_tuple_operands.begin(), old_return_tuple_operands.end());
+  new_return_tuple_operands.push_back(add);
+  HloInstruction* new_return_tuple = while_body->AddInstruction(
+      HloInstruction::CreateTuple(new_return_tuple_operands));
+  while_body->set_root_instruction(new_return_tuple, true);
+  return gte;
+}
+
+// This function takes an array literal and gives a constant instruction with
+// that literal.
+template <typename T>
+std::unique_ptr<HloInstruction> GetAsConstantInstruction(
+    const std::vector<Literal>& offset_values, PrimitiveType ty) {
+  if (ShapeUtil::IsScalarWithElementType(offset_values[0].shape(), ty)) {
+    Array<T> constantLiterals({(int64_t)offset_values.size()});
+    std::vector<T> valuesAsTy;
+    valuesAsTy.reserve(offset_values.size());
+    for (auto& i : offset_values) {
+      valuesAsTy.push_back(static_cast<T>(i.data<T>()[0]));
+    }
+    constantLiterals.SetValues(valuesAsTy);
+    return HloInstruction::CreateConstant(
+        LiteralUtil::CreateFromArray(constantLiterals));
   }
-  return false;
+  return nullptr;
 }
 
-// This checks whether a dynamic index operation has all offsets that are either
-// constant or loop iteration offsets.
-bool HasConstantOrLoopIterationOffsets(
-    const HloDynamicIndexInstruction& instr) {
-  return llvm::all_of(instr.index_operands(), [](const HloInstruction* offset) {
-    return IsLoopIterationNumber(*offset) ||
-           IsHandledConstantForDynamicSliceFusion(*offset);
-  });
+// This function takes an array literal and gives a constant instruction with
+// that literal.
+std::unique_ptr<HloInstruction> GetAsConstantInstruction(
```

Comment:
I think this can be simplified using stuff from primitive_util.h:

https://github.com/openxla/xla/blob/c4ae62b9e470fda4809f798872a20a707cd18653/xla/primitive_util.h#L496

```using NativeT = typename primitive_util::PrimitiveTypeToNative<type>::type;```

Reply:
I couldn't use the `PrimitiveTypeToNative` - it requires the template argument to be a constant, and we would like that argument to be `offset_values[0].shape().element_type()`. But I used `NativeToPrimitiveType`. Let me know if you have something nicer in mind. Thanks!

Reply:
It only works in combination with PrimitiveTypeSwitch. This will forward the different possible PrimitiveType values as constants to your templated function. Sorry if that wasn't clear.

Reply:
I think here is a pointer to existing code that should be helpful:

https://github.com/openxla/xla/blob/main/xla/service/hlo_parser.cc#L4207

Reply:
Done. Thanks for the recommendation!